src/ResourceManager/Profile/Commands.Profile/Models/RMProfileClient.cs

```diff
@@ -117,86 +118,118 @@ public bool TryRemoveContext(IAzureContext context)
                 {
-                    account.SetOrAppendProperty(AzureAccount.Property.Tenants, new[] { newTenant.Id.ToString() });
+                    newSubscription = new AzureSubscription
+                    {
+                        Id = subscriptionId,
+                    };
                 }
+
+                else
+                {
+                    newSubscription = new AzureSubscription
+                    {
+                        Name = subscriptionName
+                    };
+                }
+
+                newTenant = new AzureTenant
+                {
+                    Id = tenantId
+                };
             }
-            // (tenant is not provided and subscription is present) OR
-            // (tenant is not provided and subscription is not provided)
+
             else
             {
-                var tenants = ListAccountTenants(account, environment, password, promptBehavior, promptAction)
-                    .Select(s => s.Id.ToString()).ToArray();
-                account.SetProperty(AzureAccount.Property.Tenants, null);
-                string accountId = null;
-
-                for (int i = 0; i < tenants.Count(); i++)
+                // (tenant and subscription are present) OR
+                // (tenant is present and subscription is not provided)
+                if (!string.IsNullOrEmpty(tenantId))
+                {
+                    var token = AcquireAccessToken(
+                        account,
+                        environment,
+                        tenantId,
+                        password,
+                        promptBehavior,
+                        promptAction);
+                    if (TryGetTenantSubscription(
+                        token,
+                        account,
+                        environment,
+                        tenantId,
+                        subscriptionId,
+                        subscriptionName,
+                        out newSubscription,
+                        out newTenant))
+                    {
+                        account.SetOrAppendProperty(AzureAccount.Property.Tenants, new[] { newTenant.Id.ToString() });
+                    }
+                }
+                // (tenant is not provided and subscription is present) OR
+                // (tenant is not provided and subscription is not provided)
+                else
                 {
-                    var tenant = tenants[i];
+                    var tenants = ListAccountTenants(account, environment, password, promptBehavior, promptAction)
+                        .Select(s => s.Id.ToString()).ToArray();
+                    account.SetProperty(AzureAccount.Property.Tenants, null);
+                    string accountId = null;
 
-                    IAzureTenant tempTenant;
-                    IAzureSubscription tempSubscription;
+                    for (int i = 0; i < tenants.Count(); i++)
```

Comment:
`tenants` is an array, so you should use `tenants.Length`. :)

Any time you have a concrete `List<T>` or `T[]`, you should use the O(1) operations on the concrete type.  For arrays, it is `.Length`, for lists, it is `.Count`.

However, since arrays and lists are both enumerable, they get all of the `IEnumberable` methods, which is why you will see `IEnumerable<T>.Count()` in the intellisense.  All methods on `IEnumerable` will re-enumerate the collection, so `.Count()` is an O(n) operation.

Also, I would actually do `.ToList()` for `tenants`.  `List<T>`s are [almost alwyas](https://stackoverflow.com/questions/434761/array-versus-listt-when-to-use-which) preferred over `T[]`s.  It's not a big deal either way, but it's just kind of a C# standard practice, at this point, to use lists unless an array is specifically required.

Lastly, you might prefer a `foreach` here, and I will make a note of how to address the `i` comparison below.

Reply:
@twitchax Actually none of this code is mine, I just moved it into an else block to accommodate the -SkipValidation parameter. I'm happy to refactor though.

Reply:
Done